### PR TITLE
Cholesky Upper Factorization

### PR DESCRIPTION
**Pull request by Sely85**
`Upper` version of Cholesky factorization: both local and distributed implementation.

**Notes**
- `Panel` has been used for the distributed implementation.
- Tests work for GPUs as well.

### CODE REVIEW

File: test/unit/test_lapack_tile/test_potrf.h [hunk 1]
```diff
@@ -29,11 +29,10 @@ using namespace testing;
 
 template <class T, Device D, bool return_info>
 void testPotrf(const blas::Uplo uplo, const SizeType n, const SizeType extra_lda) {
+  std::function<T(const TileElementIndex&)> el_a, res_a;
```
Comment (albestro) [inline test/unit/test_lapack_tile/test_potrf.h:32]: minor thing: why do you prefer it here? IMHO it is better to have it near its initialization.

File: include/dlaf/factorization/cholesky/impl.h [hunk 2]
```diff
@@ -112,6 +138,8 @@ void Cholesky<backend, device, T>::call_L(comm::CommunicatorGrid grid, Matrix<T,
   auto executor_np = dlaf::getNpExecutor<backend>();
   auto executor_mpi = dlaf::getMPIExecutor<backend>();
 
+  auto lower = blas::Uplo::Lower;
+
```
Comment (albestro) [inline include/dlaf/factorization/cholesky/impl.h:142]: it seems it is not used

File: include/dlaf/factorization/cholesky/impl.h [hunk 3]
```diff
@@ -68,6 +67,33 @@ void gemmTrailingMatrixTile(Executor&& trailing_matrix_executor,
...
+template <class Executor, Device device, class T>
+void gemmTrailingMatrixTileU(Executor&& trailing_matrix_executor,
+                             hpx::shared_future<matrix::Tile<const T, device>> panel_tile,
+                             hpx::shared_future<matrix::Tile<const T, device>> col_panel,
+                             hpx::future<matrix::Tile<T, device>> matrix_tile) {
+  hpx::dataflow(trailing_matrix_executor, matrix::unwrapExtendTiles(tile::gemm_o), blas::Op::ConjTrans,
+                blas::Op::NoTrans, T(-1.0), std::move(panel_tile), std::move(col_panel), T(1.0),
+                std::move(matrix_tile));
+}
+
```
Comment (rasolca) [inline include/dlaf/factorization/cholesky/impl.h:96]: As different algorithm in the same namespace might might need helper functions with the same name, which are not fully equivalent, I suggest to embed them in the namespaces
`dlaf::factorization::internal::cholesky_L` and `dlaf::factorization::internal::cholesky_U`
similarly to what has been done in #392 for the triangular solver:
https://github.com/eth-cscs/DLA-Future/blob/37e4c461a0f83011f384b2d3892a79e287add2f4/include/dlaf/solver/triangular/impl.h#L36-L55
https://github.com/eth-cscs/DLA-Future/blob/37e4c461a0f83011f384b2d3892a79e287add2f4/include/dlaf/solver/triangular/impl.h#L60

File: test/include/dlaf_test/matrix/util_generic_blas.h [hunk 4]
```diff
@@ -105,7 +105,7 @@ auto getLeftTriangularSystem(blas::Uplo uplo, blas::Op op, blas::Diag diag, T al
 ///      = (X_ij * op(A)_jj + (kk-1) * gamma) / alpha,
 /// where gamma = (j+1) / (i+2) * exp(I*(i+2*j)),
 ///       kk = j+1 if op(a) is an upper triangular matrix, or
-///       kk = m-j if op(a) is an upper triangular matrix.
+///       kk = m-j if op(a) is an lower triangular matrix.
```
Comment (rasolca) [inline test/include/dlaf_test/matrix/util_generic_blas.h:108]:
```suggestion
///       kk = n-j if op(a) is an lower triangular matrix.
```
And in line 90 it should be: `A(n x n)`

**Comment by rasolca** [comment]
bors try

**Comment by Sely85** [comment]
bors try

**Review by rasolca — APPROVED** [review]

**Pull request merged**